### PR TITLE
FEAT: Add startswith and endswith

### LINKED ISSUE #2790: Add STARTS_WITH and ENDS_WITH
In ibis-project/ibis-bigquery#61 , I suggested to add the big query functions `STARTS_WITH` and `ENDS_WITH` to the BigQuery repo. @tswast, rightfully, pointed out that this could go into the main repo. 

Should I go ahead and make a PR? Any thing I should look out for?

I understand the operation definitions belong in _operations.py_. Would a PR also require an implementation for each backend?

My current implementation looks for BigQuery looks like this:

```python
import ibis.expr.rules as rlz
from ibis.expr.operations import Arg, ValueOp
from ibis.expr.types import StringValue
from ibis_bigquery import BigQueryExprTranslator

compiles = BigQueryExprTranslator.compiles


class StartsWith(ValueOp):
    value = Arg(rlz.string)
    start_string = Arg(rlz.string)
    output_type = rlz.shape_like("value", "bool")


def startswith(value, start_string):
    return StartsWith(value, start_string).to_expr()


class EndsWith(ValueOp):
    value = Arg(rlz.string)
    end_string = Arg(rlz.string)
    output_type = rlz.shape_like("value", "bool")


def endswith(value, start_string):
    return EndsWith(value, start_string).to_expr()


StringValue.startswith = startswith
StringValue.endswith = endswith


@compiles(StartsWith)
def _startswith(t, expr):
    # pull out the arguments to the expression
    value, start_string = expr.op().args
    # compile the argument
    t_value = t.translate(value)
    t_start = t.translate(start_string)
    # return a SQL expression that calls the BigQuery STARTS_WITH function
    return f"STARTS_WITH({t_value}, {t_start})"


@compiles(EndsWith)
def _endswith(t, expr):
    # pull out the arguments to the expression
    value, start_string = expr.op().args
    # compile the argument
    t_value = t.translate(value)
    t_start = t.translate(start_string)
    # return a SQL expression that calls the BigQuery STARTS_WITH function
    return f"ENDS_WITH({t_value}, {t_start})"

```

### PR DESCRIPTION
**Pull request by saschahofmann**
Closes #2790 .

Adds `StartsWith` and `EndsWith` operations and adds compilers to base SQL and SQLAlchemy.

I haven't written any tests yet. 

I assume I should add the operations to `test_ops_smoke` in *expr/test_operations.py* and test them in _sql/test_compiler.py_ and _sql/test_sqlalchemy.py_ ?

Anything else I am missing?

### CODE REVIEW

**Comment by saschahofmann** [comment]
I added some tests, still working on the sqlalchemy tests.

As a general comment, I found that adding tests is pretty unintuitive. A lot of different approaches and some confusing stuff. e.g. why are the SQLAlchemy tests overwriting the `alltypes` table to be `functional_alltypes`? 

Not an expert but relying on children props in parent class methods seems like an anti-pattern (looking at `self.con` in `ExprTestCases`). I guess cleaning up the test suite is a huuuuge undertaking though 😅

**Comment by saschahofmann** [comment]
Ok I added the sqlalchemy test but it fails. Anyone knows how I would make both queries return the same param names?

Test implementation:
```python
    def test_startswith(self):
        t = self.alltypes
        sat = self.sa_alltypes.alias('t0')

        expr = t.string_col.startswith('foo')
        sa_expr = sa.select([sat.c.string_col.startswith('foo').label('tmp')])

        assert self._compare_sqla(expr, sa_expr)
 ```

The ibis repr is
```SELECT (t0.string_col LIKE :string_col_1 || '%') AS tmp
FROM functional_alltypes AS to
```
And SQLAlchemy:
```SELECT (t0.string_col LIKE :param_1 || '%') AS tmp
FROM functional_alltypes AS to
```

So the only difference is in the param names (`string_col_1` and `param_1`). 


Do I need to test this in a backend as well?

File: ibis/backends/base/sql/registry/string.py [hunk 1]
```diff
@@ -80,3 +80,21 @@ def parse_url(translator, expr):
...
         )
+
+
+def startswith(translator, expr):
+    arg, start = expr.op().args
+
+    arg_formatted = translator.translate(arg)
+    start_formatted = translator.translate(start)
+
+    return f"starts_with({arg_formatted}, {start_formatted})"
```
Comment (datapythonista) [inline ibis/backends/base/sql/registry/string.py:91]: I don't think any backend other than bigquery supports this, use `LIKE 'XXX%'` instead, which should work on all them.

File: ibis/backends/base/sql/registry/string.py [hunk 2]
```diff
@@ -80,3 +80,21 @@ def parse_url(translator, expr):
...
+
+
+def endswith(translator, expr):
+    arg, start = expr.op().args
+
+    arg_formatted = translator.translate(arg)
+    end_formatted = translator.translate(start)
+
+    return f"ends_with({arg_formatted}, {end_formatted})"
```
Comment (datapythonista) [inline ibis/backends/base/sql/registry/string.py:100]: same

**Review by datapythonista — COMMENTED** [review]
I wish only the tests needed refactoring. ;)

I think there should be tests in `ibis/backends/tests` which are then executed for all backends (including third-party backends like bigquery). Not sure about the difference with sqlalchemy. Look for `like` or a similar function, and see how it's being tested. Hopefully it's tested well enough, and can be used as a reference.

File: ibis/expr/api.py [hunk 3]
```diff
@@ -2888,6 +2888,48 @@ def _string_join(self, strings):
...
+    Parameters
+    ----------
+    start: string
+
+    Examples
+    --------
+    >>> import ibis
+    >>> text = ibis.literal('Ibis project)
+    >>> result = text.startswith('Ibis')
+
```
Comment (jreback) [inline ibis/expr/api.py:2904]: can you add some examples here (and in endswith)

**Review by jreback — CHANGES_REQUESTED** [review]
this is trivial to implement on the pandas backend if you can. also pls add a release note.

File: ibis/expr/api.py (hunk 3 above)
Comment (saschahofmann) [reply ibis/expr/api.py:2904]: You mean I should mention the result? Because above is showing an example on how to use or?

**Comment by saschahofmann** [comment]
Looks like None of the similar string functions are tested in _test_sqlalchemy.py_. So I will leave it for now, let me know if I should try to include this there. Anyway, it seems to work just the parameter name is different.

**Comment by saschahofmann** [comment]
Added the functions to the backend test and to the pandas compiler. I think I also need to change something in the dask backend but let's see where else the tests fail.

File: ibis/expr/api.py (hunk 3 above)
Comment (datapythonista) [reply ibis/expr/api.py:2904]: Yes, you should write the output. Not sure if we're running these doctests at the moment (we do run some), but they'll fail if/when we do.

File: ibis/backends/tests/test_string.py [hunk 4]
```diff
@@ -171,6 +171,10 @@ def test_string_col_is_unicode(backend, alltypes, df):
             id='length',
             marks=pytest.mark.xfail_backends(['omniscidb']),  # #2338
         ),
+        param(lambda t: t.string_col.startswith('foo'),
+              lambda t: t.string_col.startswith('foo'), id='startswith'),
+        param(lambda t: t.string_col.endswith('foo'),
+              lambda t: t.string_col.endswith('foo'), id='endsswith'),
```
Comment (datapythonista) [inline ibis/backends/tests/test_string.py:177]: This should be testing the SQLAlchemy implementation. If you want to be sure, you can break something on the SQLAlchemy function, and then run `PYTEST_BACKENDS="sqlite" python -m pytest ibis/backends/tests` and see if this test fails.

Comment (saschahofmann) [reply ibis/backends/tests/test_string.py:177]: Looks like the second function is working on a pandas DataFrame though?

Comment (saschahofmann) [reply ibis/backends/tests/test_string.py:177]: Is there a setup script to setup the databases? + I am getting errors of CSV files missing.

**Comment by saschahofmann** [comment]
Also looks like the black formatter is not in sync with the pep8 bot. I actually had to overwrite my on save changes from vscode to satisfy the bot but now the CI failed on the exact same line 😅 
 Is the formatted usually failing? Don't see what I changed in that file or should I merge these changes:

image

File: ibis/expr/api.py (hunk 3 above)
Comment (saschahofmann) [reply ibis/expr/api.py]: Ok I updated this! Hope that's what you expected?

**Comment by saschahofmann** [comment]
I made a silly mistake in the string compiler so hopefully, the impala and SQL tests pass now. I also added a pyspark implementation but I haven't used pyspark yet so might be worth checking that? 

I haven't managed to run the backend tests locally yet (see discussion above) but maybe someone could trigger the workflow again 🤓 .

**Comment by saschahofmann** [comment]
Is it possible that impala uses `||` as or operator rather than string concatenation? the error in the failing test suggests that much

File: ibis/expr/api.py (hunk 3 above)
Comment (saschahofmann) [reply ibis/expr/api.py]: @jreback are you happy with my changes to the example? Or did you expect a different format. Didnt really see many examples for this

**Comment by datapythonista** [comment]
> Is it possible that impala uses `||` as or operator rather than string concatenation? the error in the failing test suggests that much

I don't understand where the test failure is coming from, but Impala uses `OR`, as in standard SQL: https://impala.apache.org/docs/build/html/topics/impala_operators.html#ariaid-title13

And afaik, in SQL string concatenation is usually not done with an operator, but with a function. That seems to be the case for Impala: https://impala.apache.org/docs/build/html/topics/impala_string_functions.html#string_functions__concat

I hope this helps.

**Comment by saschahofmann** [comment]
I tried to find something about this in the Impala docs but only found this (contrary to how I understand the test failure):
> Logical/Boolean operators AND, OR, and NOT. Impala does not support the corresponding symbols &&, ||, and !.

If I use a function, I'd need to make sure it's the same for all maybe I just implement a separate case for Impala?

**Comment by saschahofmann** [comment]
Nvm I fixed another bug and now I am using the contact operator. If you could trigger the test suite again 🤞

**Comment by datapythonista** [comment]
There is an issue with black. The other errors in the CI are unrelated to this PR.

**Comment by saschahofmann** [comment]
Ok should be fixed

**Comment by saschahofmann** [comment]
Ok I think this is finally ready? 

Sorry for taking so long 😏

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @saschahofmann very nice (failures are unrelated).

**Pull request merged**